promql/engine.go

```diff
@@ -697,117 +655,355 @@ func (ev *evaluator) eval(expr Expr) Value {
+			}
+			return ev.aggregation(e.Op, e.Grouping, e.Without, param, v[1].(Vector), enh)
+		}, e.Param, e.Expr)
 
-	case *BinaryExpr:
-		lhs := ev.evalOneOf(e.LHS, ValueTypeScalar, ValueTypeVector)
-		rhs := ev.evalOneOf(e.RHS, ValueTypeScalar, ValueTypeVector)
+	case *Call:
+		if e.Func.Name == "timestamp" {
+			// Matrix evaluation always returns the evaluation time,
+			// so this function needs special handling when given
+			// a vector selector.
+			vs, ok := e.Args[0].(*VectorSelector)
+			if ok {
+				return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+					return e.Func.Call([]Value{ev.vectorSelector(vs, enh.ts)}, e.Args, enh)
+				})
+			}
+		}
+		// Check if the function has a matrix argument.
+		for matrixArgIndex, a := range e.Args {
+			sel, ok := a.(*MatrixSelector)
+			if ok {
+				inArgs := make([]Value, len(e.Args))
+				// Evaluate any non-matrix arguments.
+				otherArgs := make([]Matrix, len(e.Args))
+				otherArgsIn := make([]Vector, len(e.Args))
+				for i, e := range e.Args {
+					if i != matrixArgIndex {
+						otherArgs[i] = ev.eval(e).(Matrix)
+						otherArgsIn[i] = Vector{Sample{}}
+						inArgs[i] = otherArgsIn[i]
+					}
+				}
 
-		switch lt, rt := lhs.Type(), rhs.Type(); {
-		case lt == ValueTypeScalar && rt == ValueTypeScalar:
-			return Scalar{
-				V: scalarBinop(e.Op, lhs.(Scalar).V, rhs.(Scalar).V),
-				T: ev.Timestamp,
+				mat := make(Matrix, 0, len(sel.series)) // Output matrix.
+				offset := durationMilliseconds(sel.Offset)
+				// Reuse objects across steps to save memory allocations.
+				points := getPointSlice(16)
+				inMatrix := make(Matrix, 1)
+				inArgs[matrixArgIndex] = inMatrix
+				enh := &evalNodeHelper{out: make(Vector, 0, 1)}
+				// Process all the calls for one time series at a time.
+				var it *storage.BufferedSeriesIterator
+				for i, s := range sel.series {
+					if it == nil {
+						it = storage.NewBuffer(s.Iterator(), durationMilliseconds(sel.Range))
+					} else {
+						it.Reset(s.Iterator())
+					}
+					ss := Series{
+						// For all range vector functions, the only change to the
+						// output labels is dropping the metric name so just do
+						// it once here.
+						Metric: dropMetricName(sel.series[i].Labels()),
+						Points: getPointSlice(numSteps),
+					}
+					inMatrix[0].Metric = sel.series[i].Labels()
+					for ts, step := ev.Timestamp, -1; ts <= ev.EndTimestamp; ts += ev.Interval {
+						step++
+						// Set the non-matrix arguments.
+						// They are scalar, so it is safe to use the step number
+						// when looking up the argument, as there will be no gaps.
+						for j := range e.Args {
+							if j != matrixArgIndex {
+								otherArgsIn[j][0].V = otherArgs[j][0].Points[step].V
+							}
+						}
+						maxt := ts - offset
+						mint := maxt - durationMilliseconds(sel.Range)
+						// Evaluate the matrix selector for this series for this step.
+						points = ev.matrixIterSlice(it, maxt, mint, points[:0])
```

Comment:
Would be worth checking if there is any overlap between previous and desired points (as would be the case with e.g. `rate[5m]` with an interval of `1m`) copying over the overlap and requesting only the difference.

Might even integrate that with the iterator and only `Seek` once.

Probably better done in a separate PR though, this one is already huge.

Reply:
Well, I have no idea how Go handles memory allocations, but I would guess it's more efficient to `copy()` the pointers to overlapping `Point` instances than to create new ones.

Particularly with very long ranges. I always kick myself when I change the range in a query but not the resolution and everything just hangs for a while.

Reply:
Most of this PR is about reducing memory allocations. Given how things work you want to copy Points structs around, otherwise all those Point objects would need to be GCed.

> I always kick myself when I change the range in a query but not the resolution and everything just hangs for a while.

That'll be a lot better after this PR :)

Reply:
I realize now I was thinking in Java (it's a difficult context switch to Go) and was assuming these were actually `Point` pointer arrays rather than `Point` struct arrays.

Even with struct arrays, it's still likely that a `copy()` will be more efficient than repopulating the array one by one, even from memory, but one could argue it is not worth the added complexity.